psi4/src/psi4/dlpno/mp2.cc

```diff
@@ -0,0 +1,1412 @@
+                                            T_iajb[ij]->get(a, b));
+                }
+            }
+
+            for (int k = 0; k < naocc; ++k) {
+                int kj = i_j_to_ij[k][j];
+                int ik = i_j_to_ij[i][k];
+
+                if (kj != -1 && i != k && fabs(F_lmo->get(i, k)) > options_.get_double("F_CUT") && n_pno[kj] > 0) {
+                    SharedMatrix temp =
+                        linalg::triplet(S_pno_ij_kj[ij][k], T_iajb[kj], S_pno_ij_kj[ij][k], false, false, true);
+                    temp->scale(-1.0 * F_lmo->get(i, k));
+                    R_iajb[ij]->add(temp);
+                }
+                if (ik != -1 && j != k && fabs(F_lmo->get(k, j)) > options_.get_double("F_CUT") && n_pno[ik] > 0) {
+                    SharedMatrix temp =
+                        linalg::triplet(S_pno_ij_ik[ij][k], T_iajb[ik], S_pno_ij_ik[ij][k], false, false, true);
+                    temp->scale(-1.0 * F_lmo->get(k, j));
+                    R_iajb[ij]->add(temp);
+                }
+            }
+
+            R_iajb_rms[ij] = R_iajb[ij]->rms();
+        }
+
+        // evaluate convergence using current amplitudes and residuals
+        e_prev = e_curr;
+        e_curr = eval_amplitudes();
+        r_curr = *max_element(R_iajb_rms.begin(), R_iajb_rms.end());
+
+        r_converged = (fabs(r_curr) < options_.get_double("R_CONVERGENCE"));
+        e_converged = (fabs(e_curr - e_prev) < options_.get_double("E_CONVERGENCE"));
+
+// use residuals to get next amplitudes
+#pragma omp parallel for schedule(static, 1)
+        for (int ij = 0; ij < n_lmo_pairs; ++ij) {
+            int i, j;
+            std::tie(i, j) = ij_to_i_j[ij];
+            for (int a = 0; a < n_pno[ij]; ++a) {
+                for (int b = 0; b < n_pno[ij]; ++b) {
+                    T_iajb[ij]->add(a, b, -R_iajb[ij]->get(a, b) / ((e_pno[ij]->get(a) + e_pno[ij]->get(b)) -
+                                                                    (F_lmo->get(i, i) + F_lmo->get(j, j))));
+                }
+            }
+        }
+
+        // DIIS extrapolation
+        SharedMatrix T_iajb_flat = flatten_mats(T_iajb);
+        SharedMatrix R_iajb_flat = flatten_mats(R_iajb);
+        if (iteration == 0) {
+            diis->set_error_vector_size(1, DIISEntry::Matrix, R_iajb_flat.get());
+            diis->set_vector_size(1, DIISEntry::Matrix, T_iajb_flat.get());
+        }
+        diis->add_entry(R_iajb_flat, T_iajb_flat);
+        diis->extrapolate(T_iajb_flat);
+        copy_flat_mats(T_iajb_flat, T_iajb);
+
+#pragma omp parallel for schedule(static, 1)
+        for (int ij = 0; ij < n_lmo_pairs; ++ij) {
+            Tt_iajb[ij]->zero();
+            Tt_iajb[ij]->add(T_iajb[ij]);
+            Tt_iajb[ij]->scale(2.0);
+            Tt_iajb[ij]->subtract(T_iajb[ij]->transpose());
+        }
+
+        outfile->Printf("  @LMP2 iter %3d: %16.12f %10.3e %10.3e\n", iteration, e_curr, e_curr - e_prev, r_curr);
+
+        iteration++;
```

Comment:
Can we have a max iterations check? By all accounts, iterative MP2 should converge very fast. But just in case something goes terribly wrong, we should terminate the computation.

Reply:
Sure, I added a `DLPNO_MAXITER` variable and iteration check